### PR TITLE
CI templates

### PR DESCRIPTION
**Pull request by aminya**
### Description of the change
- This uses Powershell 7 as a cross-platform shell for CI scripts:
	- PowerShell 7 allows using the same syntax and code across all the operating systems.
	- This simplifies CI scripts. For example, previously CMD was used in Windows, which is an old shell with limited functionality.

- This refactors CI scripts into separate files (templates): 
	- It also allows using parallelization of the tests without copy-pasting the same code. For example: 
		- In macOS, the same templates are reused in the test phase. 
		- In Windows x86, the templates are used for parallelization of tests in #31  

	- This makes it easier to manage CI code. For example, we can quickly compare the publish step for all operating systems without going through big files and trying to find the relevant steps.

Previously, CI scripts were written back in ~2017. At that time, Azure Templates had not been introduced yet, and because of that using PowerShell was not suitable yet. Now in 2020, we can benefit from these and improve the CI.

### Verification
The CI passes

### Release Notes
N/A

### CODE REVIEW

**Comment by aminya** [comment]
@DeeDeeG Because of this, we need to rebase other pull requests. So let's first merge this.

**Comment by DeeDeeG** [comment]
I hope we can combine the templates, especially macOS/Linux which are both UNIX-like and use mostly the same commands.

We can use [conditional expressions](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#conditional-insertion) looking at `"$(Agent.OS)"` to set these env vars, for example: https://github.com/atom-ide-community/atom/blob/91f302f5451780bf19d6d84436647297080b4ce9/script/vsts/platforms/templates/linux-bootstrap.yml#L7-L9

**Comment by DeeDeeG** [comment]
I'm personally not sure this is more readable. It's inconvenient to open multiple files and remember how they are referenced across several files.

**Comment by aminya** [comment]
The code for three operating systems is different enough (the scripts, paths, file names, etc). 

If we want to combine these, we should use the same shell (Bash or PowerShell), and find a way to parameterize the paths, and file names.

**Comment by DeeDeeG** [comment]
This all seems good in the abstract... I am working on seeing why our CI doesn't tend to pass and I am finding all this diff that we already have from upstream hard to debug. Shouldn't we pause on this sort of stuff until we can get CI reliably passing?

And thank you for working on it, reagrdless.

**Comment by aminya** [comment]
> This all seems good in the abstract... I am working on seeing why our CI doesn't tend to pass 

This and #31 solve our CI problems.

**Comment by DeeDeeG** [comment]
I mean that upstream passes and ours doesn't. I hope you are right that these changes make CI pass. I am continuing to investigate, such as with the [`investigate-CI`](https://github.com/atom-ide-community/atom/tree/investigate-CI) branch and its CI runs.

**Comment by DeeDeeG** [comment]
Okay, I think the failures here at the fork are timeouts. I think we're doing okay, just that we're less patient to re-run failed tests due to timeouts. In which case you are right, parallelizing will help us "fail faster" and "re-run faster."

I note that the upstream caching solution saved its cache right after bootstrapping, so that solution lets us skip bootstrapping the second go-around (saves ~15 minutes)... So there might actually be an advantage to reverting Cache@2 now that we know how to make SaveCache@1 work.

**Comment by aminya** [comment]
I want to merge this pull request before merging any other CI pull requests to prevent conflict issues. I merged #45 into this if you have not seen it yet. We should not revert Cache2 when it is working for us.

**Comment by DeeDeeG** [comment]
I prefer to merge #45 into `master` and then rebase this on `master`.

I'm not even sure if I like these templates. They don't cause re-runs to happen any faster, so... Unless we can convert them into separate stages or jobs, I don't think there's a real benefit.

I must say I find them harder to read than before.

If upstream changes their CI configs it could be a nightmare to rebase on top of. They just did that recently with the nightly dependency updates.

**Comment by aminya** [comment]
Just bear with me for some time 😄 I have the experience of building [pretty complex software](https://github.com/include-what-you-use/include-what-you-use/pull/799/files). Combining the steps is helpful even if it does not help readability. This is similar to C++ templates. Although they look complicated, they are the best thing to do. Parallelization needs this PR otherwise I have to copy-paste many times.

**Comment by DeeDeeG** [comment]
If we do in fact use one shell cross-platform, I would vote for Powershell. :+1:

**Comment by aminya** [comment]
Linux cannot find pwsh for some reason. I updated the container and vmImage. It seems to be related to this: https://github.com/Microsoft/azure-pipelines-agent/issues/1862

**Comment by DeeDeeG** [comment]
Can we try with no container? I'm not sure why we need to run under Docker. I would assume the Ubuntu docker container, which is fairly minimal, doesn't have Powershell.

**Comment by aminya** [comment]
> Can we try with no container? I'm not sure why we need to run under Docker. I would assume the Ubuntu docker container, which is fairly minimal, doesn't have Powershell.

Yes. Probably that is the problem! 🤷‍♂️ I will try that.

**Comment by DeeDeeG** [comment]
> sudo apt-add-repository "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-5.0 main"

This is the problem. This is a specifically trusty-oriented PPA.

Just find-replace "trusty" --> "xenial" (16.04) or "bionic" (18.04)

**Comment by aminya** [comment]
Fortunately, pwsh is found without a container. Now we just need to update the Linux dependencies. Building using recent compilers also improves performance!

**Comment by DeeDeeG** [comment]
Clang 5 is super old. Clang 10 is current stable. https://releases.llvm.org/download.html

**Comment by aminya** [comment]
We do not even need to install Clang. It is already included:

https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1604-README.md
https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md

**Comment by DeeDeeG** [comment]
True in a lot of ways about Windows too, not to get too off-topic.

**Comment by aminya** [comment]
> True in a lot of ways about Windows too, not to get too off-topic.
> 
> Just use what's included.

I agree. We will fix these all. :smile:

**Comment by aminya** [comment]
I am not going to modify the Linux dependencies here much. I just wanted the PowerShell to be detected. I will keep that for a separate PR. Similar to our #44.

**Comment by DeeDeeG** [comment]
> ```
> make: clang++-9.0: Command not found
> ```

**Comment by DeeDeeG** [comment]
We can set fetch-depth to `1` on the pipeline to save about a minute per run of this.

I have had several passing CI runs with that setting on my personal fork. If you don't mind I'd like to set that.

**Comment by aminya** [comment]
Sure. Is it inside Azure UI?

**Comment by DeeDeeG** [comment]
[Pipeline] --> [Edit] --> [Triggers] --> [left-most tab] --> [source fetch] (I think) --> fetch depth setting

**Comment by DeeDeeG** [comment]
Apt dependencies successfully installed with Clang 9! :tada:

**Comment by DeeDeeG** [comment]
I set the fetch-depth to 1. Also the instructions above are updated now that I had another look at the settings pages.

**Comment by DeeDeeG** [comment]
[Off-topic/ deleting in a moment.]

That's it for me for today, good luck. :+1:

**Comment by aminya** [comment]
Hit an issue: https://stackoverflow.com/questions/62890370/azure-pipeline-conditional-env-variable-based-on-os?noredirect=1#comment111213813_62890370

**Comment by aminya** [comment]
This is almost ready. Only win x86 script fails to run the tests. I am going to use CMD for win x86 instead of Powershell if I could not find any other solution.

**Comment by aminya** [comment]
This is ready to go 🚀

**Comment by DeeDeeG** [comment]
Thanks for all the work. I'll be taking a closer look some time at it so I understand what;s changed. Apologies for all the grumbling about this. I may have gone overboard with that.

A request: I'd love it if we could squash this into fewer commit, with clearly separated changes per commit, and containing (to the extent possible) just the final version of this PR's changes. This would help ensure we don't get confused when trying to rebase on `upstream_master` in the future. I know I personally would appreciate it, anyhow.

Something like this:
- Commit 1: Add the new templates files (don't use them yet)
- Commit 2: Use the template files in platforms definitions and pipeline definitions

(Maybe split up "using in platform definitions" (Windows, macOS, Linux) from "using templates in pipeline definitions" (nightly, pull requests,release branches... But I'm not sure there would be a benefit.)

File: script/vsts/platforms/windows.yml [hunk 1]
```diff
@@ -20,221 +20,48 @@ jobs:
...
-          ArtifactName: crash-reports
-        displayName: Publish crash reports on non-release branch
-        condition: and(failed(), eq(variables['ATOM_RELEASES_S3_KEY'], ''))
+      - template: templates/test.yml
 
-      - script: >
-          node $(Build.SourcesDirectory)\script\vsts\upload-crash-reports.js --crash-report-path "%ARTIFACT_STAGING_DIR%\crash-reports" --s3-path "vsts-artifacts/%BUILD_ID%/"
+      - pwsh: |
+          if ($env:BUILD_ARCH -eq "x64") {
+            $env:FileID="-x64"
```
Comment (DeeDeeG) [inline script/vsts/platforms/windows.yml:42]: "`FileID`" variable name: Can we make this more clear what it is used for? Call it "`Arch`" or "`WindowsArch`", or something similar?

Or even "`Dashx64`"?

**Comment by aminya** [comment]
> As mentioned above, this would be MUCH EASIER to rebase in the event of a merge conflict. 

I decided to not rebase our master anymore. See my comment here https://github.com/atom-ide-community/atom/issues/28#issuecomment-659764321. It is harmful to our repository.

File: script/vsts/platforms/templates/download-unzip.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+steps:
+  - ${{ each artifact in parameters.artifacts }}:
+    - task: DownloadBuildArtifacts@0
+      displayName: Download ${{artifact}}
+      inputs:
+        artifactName: ${{artifact}}
+        downloadPath: ${{parameters.downloadPath}}
+
+    - script: unzip ${{artifact}}/${{artifact}} -d out
+      displayName: Unzip ${{artifact}}
```
[the remark this answers is not in the document]
Comment (aminya) [reply script/vsts/platforms/templates/download-unzip.yml:18]: If you had given this sooner I could change this. Now it is a little late as this is already merged into master and changing it now will break other open pull requests. 🤔

File: script/vsts/platforms/windows.yml [hunk 3]
```diff
@@ -20,221 +20,48 @@ jobs:
...
-          NPM_BIN_PATH: "C:\\hostedtoolcache\\windows\\node\\12.13.1\\x64\\npm.cmd"
-          npm_config_build_from_source: true
-        displayName: Bootstrap build environment
-        condition: or(ne(variables['MainNodeModulesRestored'], 'true'), ne(variables['ScriptNodeModulesRestored'], 'true'), ne(variables['ApmNodeModulesRestored'], 'true'))
+      - template: templates/bootstrap.yml
 
       - script: node script\vsts\windows-run.js script\lint.cmd
         env:
           BUILD_ARCH: $(buildArch)
         displayName: Run linter
```
[the remark this answers is not in the document]
Comment (aminya) [reply script/vsts/platforms/windows.yml:34]: Linter was pretty small, so I did not bother to move it to a template. I plan to make the linter independent of the bootstrap. People should not wait 10 min to realize that they made a linting mistake.

So I will keep this for now until we fix this in https://github.com/atom-ide-community/atom/projects/6#card-41610711

File: script/vsts/platforms/windows.yml (hunk 1 above)
Comment (aminya) [reply script/vsts/platforms/windows.yml:42]: Again, this requires rebasing and changing the commit history. If you had given me a notice earlier we could do this.

**Comment by aminya** [comment]
> It feels like it could be good to have `GetReleaseVersion` in a template, too.
> 
> It's mostly identical everywhere it's used.
> 
> If you want I can try to do this.

We can do that. We run the following two times in each script. 
```
      - script: |
          cd script\vsts
          npm install
        env:
          GITHUB_TOKEN: $(GITHUB_TOKEN)
        displayName: npm install
```

So if you can make a template that includes both the following steps and have a condition to not run the `get-release-version` step, we can do that.
```
      - script: |
          cd script\vsts
          npm install
        env:
          GITHUB_TOKEN: $(GITHUB_TOKEN)
        displayName: npm install

      - script: node script\vsts\get-release-version.js
        name: Version
        env:
          REPO_OWNER: $(REPO_OWNER)
          NIGHTLY_RELEASE_REPO: $(NIGHTLY_RELEASE_REPO)

```

Please make your pull request to this branch, so I can upstream it as part of the PR.

File: script/vsts/platforms/windows.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (DeeDeeG) [reply script/vsts/platforms/windows.yml:34]: I'm not sure if they have what they need upstream. Looks like they are concerned that the linter dependency will be unnecessarily bundled in the built Atom packages.

We should have a `job` that installs the bare minimum linter tools and then runs the lints in parallel with win/mac/Linux pretty easily, though. If we want to bother with that.

(Hypothetically; I am not going to jump into this and fix it right this second).

[the remark this answers is not in the document]
Comment (aminya) [reply script/vsts/platforms/windows.yml:34]: > We should [typo -->] could have a job that installs the bare minimum linter tools and then runs the lints in parallel with win/mac/Linux pretty easily, though. If we want to bother with that.

Yes, that is what I am planning on.

File: script/vsts/platforms/templates/bootstrap.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      }
+      else {
+        script/bootstrap
+      }
+    displayName: Bootstrap build environment
+    env:
+      GITHUB_TOKEN: $(GITHUB_TOKEN)
+      CI: true
+      CI_PROVIDER: VSTS
+      condition: or(ne(variables['MainNodeModulesRestored'], 'true'), ne(variables['ScriptNodeModulesRestored'], 'true'), ne(variables['ApmNodeModulesRestored'], 'true'))
```
Comment (DeeDeeG) [inline script/vsts/platforms/templates/bootstrap.yml:31]: This indentation makes `condition` and env variable. Bootstrapping always runs.

We should indent it two spaces less, toward the left.

Comment (aminya) [reply script/vsts/platforms/templates/bootstrap.yml:31]: Thanks for catching this. I will fix it!

**Comment by DeeDeeG** [comment]
I can post updates to this branch and then open another PR from this `CI_templates` branch, and you can merge the new commits into `master` with a traditional merge commit, no rebasing needed. Sounds good?

(Then everything is in one place on this branch when you go to make a PR to upstream.)

**Comment by aminya** [comment]
I will use git cherry pick to bring your commits to master.

**Comment by trev9995** [comment]
Hey

**Pull request merged**